### ISSUE TITLE
Drift generator sub-query parameter type confusion

### ISSUE THREAD

**Issue by FaFre** [simolus3/moor#2350]
The parameter types of sub-queries are getting confused. 

Following example:
```sql
CREATE TABLE t (
  a REAL,
  b INTEGER
);

failQuery:
  SELECT 
    *, 
    LIST(SELECT * FROM t x WHERE x.b = b or x.b = :inB) 
  FROM
    (SELECT * FROM t where a = :inA AND b = :inB);
```

Produces following:
```dart
electable<FailQueryResult> failQuery({double? inA, int? inB}) {
    return customSelect(
        'SELECT * FROM (SELECT * FROM t WHERE a = ?1 AND b = ?2)',
        variables: [
          Variable<double>(inA),
          Variable<int>(inB)
        ],
        readsFrom: {
          t,
        }).asyncMap((QueryRow row) async {
      return FailQueryResult(
        a: row.readNullable<double>('a'),
        b: row.readNullable<int>('b'),
        nestedQuery0: await customSelect(
            'SELECT * FROM t AS x WHERE x.b = b OR x.b = ?1',
            variables: [
              Variable<double>(inB)
            ],
            readsFrom: {
              t,
            }).asyncMap(t.mapFromRow).get(),
      );
    });
  }
```
Suddenly the generated code wants to cast `inB` into an `double`.

Another slightly different query:
```sql
failQuery:
  SELECT *, LIST(SELECT * FROM t x WHERE x.b = 0 OR x.b = :inB) FROM t where a = :inA AND b = :inB;
```

Produces following:
```dart
Selectable<FailQueryResult> failQuery({int? inA, int? inB}) {
    return customSelect('SELECT * FROM t WHERE a = ?1 AND b = ?2', variables: [
      Variable<int>(inA),
      Variable<int>(inB)
    ], readsFrom: {
      t,
    }).asyncMap((QueryRow row) async {
      return FailQueryResult(
        a: row.readNullable<double>('a'),
        b: row.readNullable<int>('b'),
        nestedQuery0: await customSelect(
            'SELECT * FROM t AS x WHERE x.b = 0 OR x.b = ?1',
            variables: [
              Variable<int>(inB)
            ],
            readsFrom: {
              t,
            }).asyncMap(t.mapFromRow).get(),
      );
    });
  }
```
Suddenly `inA` wich is a `double` becomes an `int` in the signature.

Unfortunately I was not able to debug this behavior.

**Comment by simolus3** [comment]
Thanks for the report! The problem probably comes from here:

https://github.com/simolus3/drift/blob/fadc30df1dba4046ab70d0d56f1b8104dcac318c/sqlparser/lib/src/analysis/steps/prepare_ast.dart#L240-L252

For `LIST()` queries, we start assigning variable indices again. So `:inB` has the index 1 inside the nested query, but 2 in the main query. That is confusing the type inference algorithm, which now wants the inner `:inB` to have the same type as `:inA` since the variables have the same index.

I wonder if restarting to count indices is really necessary, I'll have to play around with this some more and see what breaks :D

**Comment by simolus3** [comment]
Should be fixed in ba86f1aa8932f4439ff5f330f0143e77591ee401.
